### PR TITLE
Sparsified SGD

### PR DESCRIPTION
**Pull request by negar-foroutan**
This is the implementation of sparsified sgd, learning rate schedules for both sgd and sparsified sgd and data loader for datasets with a lmdb format. 
Here the aggregation of gradients/parameters is included in optimizer (step function).

### CODE REVIEW

File: mlbench_core/dataset/linearmodels/pytorch/dataloader.py [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+        """
+        block_index, item_index = self._get_matched_index(index)
+        image, target = self.dbs[block_index][item_index]
+        return image, target
+
+    def __len__(self):
+        return self.length
+
+    def __repr__(self):
+        fmt_str = 'Dataset ' + self.__class__.__name__ + '\n'
```
Comment (Panaetius) [inline mlbench_core/dataset/linearmodels/pytorch/dataloader.py]: maybe use an implicitely joined string instead of the addition to make the code more readable, with a single `.format()` call
```
string = ("this is an \n"
          "implicitly joined \n"
          "string")
```

File: mlbench_core/controlflow/pytorch/controlflow.py [hunk 2]
```diff
@@ -168,7 +168,7 @@ def run(self, dataloader_train=None, dataloader_val=None,
 
                 self._get_dataloader_stats(dataloader_train, dataloader_val)
 
-    def train_epoch(self, dataloader):
+    def train_epoch(self, dataloader, dataloader_val):
```
[the remark this answers is not in the document]
Comment (negar-foroutan) [reply mlbench_core/controlflow/pytorch/controlflow.py]: I forgot to change it.

File: mlbench_core/dataset/linearmodels/pytorch/dataloader.py [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+            if len(list(from_to_indices)) == 0:
+                return 0, x
+
+            for ind, (from_index, to_index) in from_to_indices:
+                if from_index <= x and x < to_index:
+                    return ind, x - from_index
+
+        return f
+
+    def _get_matched_index(self, index):
```
[the remark this answers is not in the document]
Comment (negar-foroutan) [reply mlbench_core/dataset/linearmodels/pytorch/dataloader.py]: It's not a default pytorch method. 
I changed the code to just use _get_index_zones.
